In [1]:
#Import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
#load data from .tsv file, had to ignore an error, imported as str for ease and will alter later

ks_data = pd.read_csv('38050-0001-Data.tsv', sep='\t', dtype=str, encoding_errors='ignore')


# My Functions

In [30]:
#get portion of data
def df_head(dataframe):
    if not dataframe.head().empty:
        return dataframe.head()
    else:
        return "Your dataframe does not exist! Check your dataframe variable."

#get information
def df_info(dataframe):
    if not dataframe.empty:
        return dataframe.info() 
    else:
        return "There is an issue with your dataframe! Check your dataframe variable."

#get shape
def df_shape(dataframe):
    if dataframe.shape == (0,0):
        return "Your dataframe does not exist! Check your dataframe variable."
    else:
        return dataframe.shape

#decribe dataframe
def describe_df(dataframe):
    if not dataframe.empty:
        return dataframe.describe()
    else:
        return "Your dataframe does not exist! Check your dataframe variable."


#get columns
def get_columns(dataframe):
    list_length = len(dataframe.columns)
    if list_length >= 0:
        return dataframe.columns
    else:
        return "There are no columns in your dataframe! Check your dataframe variable."

#get column data
def column_data(dataframe, column_name: str):
    if column_name in get_columns(dataframe):
        return dataframe[column_name]
    else:
        return "Your dataframe or column does not exist! Check your dataframe variable and columns."

#check nulls
def check_nulls(dataframe):
    num_nulls = dataframe.isnull().sum()
    percent_nulls = (num_nulls / len(dataframe) * 100).round(2)
    if not dataframe.empty:
        if num_nulls.sum() > 0:
            print(f"Here are the number of nulls per column: \n{num_nulls}")
            print(f"Here are the percent of nulls per column: \n{percent_nulls}")
        else:
            return "There are no nulls in your data."
    else:
        return "Your dataframe does not exist! Check your dataframe variable."

#check value counts
def unique(dataframe):
    if not dataframe.empty:
        return f"Here are the number of unique values per columns:", dataframe.nunique()


#check for dups
def check_dups(dataframe):
    num_dups = dataframe.duplicated().sum()
    if not dataframe.empty:
        return f"There are {num_dups} duplicates in your dataframe"
    else:
        return "Your dataframe does not exist! Check your dataframe variable."
    
#drop dups
def drop_dups(dataframe):
    dataframe = dataframe.drop_duplicates()
    if check_dups(dataframe) == 0:
        return dataframe
    else:
        return "There are still duplicates in your data. Please check your code."


#Visuals 

def make_hist(dataframe, columns, num_rows, num_columns, fig_size):
    if not dataframe.empty:
       fig, axs = plt.subplots(num_rows, num_columns, figsize=fig_size)
       for i in range(num_rows):
           for j in range(num_columns):
               axs[i,j].hist(dataframe[columns[i * num_columns + j]])
               axs[i,j].set_title(columns[i * num_columns + j + 1])
    else:
        return "Your dataframe is empty. Check your dataframe variable"

#make boxplots
def make_boxplot(dataframe, columns, num_rows, num_columns, fig_size):
    if not dataframe.empty:
       fig, axs = plt.subplots(num_rows, num_columns, figsize=fig_size)
       for i in range(num_rows):
           for j in range(num_columns):
               axs[i,j].boxplot(dataframe[columns[i * num_columns + j]])
               axs[i,j].set_title(columns[i * num_columns + j + 1])
    else:
        return "Your dataframe is empty. Check your dataframe variable"

#make pairplot
def make_pairplot(dataframe):
    return sns.pairplot(dataframe)

#make heatmap
def make_heatmap(dataframe):
    correlation = dataframe.corr
    return sns.heatmap(correlation)

#make scatter plot
def make_scatter(dataframe, x_column, y_column, title: str, x_label: str, y_label: str):
    plt.scatter(x=dataframe[x_column], y=dataframe[y_column])
    plt.title(title)
    plt.xlabel(x_label)
    plt.ylabel(y_label)
    return plt.show()

# EDA

In [158]:
#make copy of raw data
ks_copy = ks_data.copy()

In [7]:
df_info(ks_copy)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506199 entries, 0 to 506198
Data columns (total 21 columns):
 #   Column                        Non-Null Count   Dtype 
---  ------                        --------------   ----- 
 0   CASEID                        506199 non-null  object
 1   NAME                          506199 non-null  object
 2   PID                           506199 non-null  object
 3   CATEGORY                      506199 non-null  object
 4   CATEGORY_ID                   506199 non-null  object
 5   SUBCATEGORY                   506199 non-null  object
 6   SUBCATEGORY_ID                506199 non-null  object
 7   PROJECT_PAGE_LOCATION_NAME    506199 non-null  object
 8   PROJECT_PAGE_LOCATION_STATE   506199 non-null  object
 9   PROJECT_PAGE_LOCATION_COUNTY  506199 non-null  object
 10  UID                           506199 non-null  object
 11  LAUNCHED_DATE                 506199 non-null  object
 12  DEADLINE_DATE                 506199 non-null  object
 13 

In [6]:
df_head(ks_copy)

CASEID             NAME         PID      CATEGORY CATEGORY_ID  \
0      1  MASKED BY ICPSR  2137925650  Film & Video          11   
1      2  MASKED BY ICPSR  1501531085  Film & Video          11   
2      3  MASKED BY ICPSR   953415668    Technology          16   
3      4  MASKED BY ICPSR  1371386304    Publishing          18   
4      5  MASKED BY ICPSR  1720842777           Art           1   

       SUBCATEGORY SUBCATEGORY_ID PROJECT_PAGE_LOCATION_NAME  \
0  Science Fiction            301                     London   
1          Fantasy            296                Los Angeles   
2         Software             51                     Mexico   
3       Publishing             18                   Columbus   
4     Illustration             22                    Toronto   

  PROJECT_PAGE_LOCATION_STATE PROJECT_PAGE_LOCATION_COUNTY  ... LAUNCHED_DATE  \
0                     England               Greater London  ...     8/11/2016   
1                          CA                  Los Angeles  ...    12/19/2019   
2             Baja California                      Tijuana  ...      3/1/2017   
3                          OH                     Franklin  ...      6/4/2018   
4                          ON                      Toronto  ...      5/1/2015   

  DEADLINE_DATE PROJECT_CURRENCY GOAL_IN_ORIGINAL_CURRENCY  \
0    10/10/2016              USD               100,000,000   
1     2/14/2020              USD               100,000,000   
2     3/22/2017              MXN               100,000,000   
3      7/5/2018              USD               100,000,000   
4     6/30/2015              CAD               100,000,000   

  PLEDGED_IN_ORIGINAL_CURRENCY   GOAL_IN_USD PLEDGED_IN_USD BACKERS_COUNT  \
0                            0  $100,000,000             $0             0   
1                           85  $100,000,000            $85             4   
2                           10    $5,219,374             $1             1   
3                            1  $100,000,000             $1             1   
4                            0   $80,610,122             $0             0   

      STATE         URL_NAME  
0  canceled  MASKED BY ICPSR  
1  canceled  MASKED BY ICPSR  
2    failed  MASKED BY ICPSR  
3  canceled  MASKED BY ICPSR  
4    failed  MASKED BY ICPSR  

[5 rows x 21 columns]

In [159]:
#I will remove some columns like Name amd URL as the data is masked and some ID columns as the data is redundant

ks_copy = ks_copy.drop(['NAME', 'PID', 'CATEGORY_ID', 'SUBCATEGORY_ID', 'UID', 'URL_NAME'], axis=1)

In [42]:
df_head(ks_copy)

CASEID      CATEGORY      SUBCATEGORY PROJECT_PAGE_LOCATION_NAME  \
0      1  Film & Video  Science Fiction                     London   
1      2  Film & Video          Fantasy                Los Angeles   
2      3    Technology         Software                     Mexico   
3      4    Publishing       Publishing                   Columbus   
4      5           Art     Illustration                    Toronto   

  PROJECT_PAGE_LOCATION_STATE PROJECT_PAGE_LOCATION_COUNTY LAUNCHED_DATE  \
0                     England               Greater London     8/11/2016   
1                          CA                  Los Angeles    12/19/2019   
2             Baja California                      Tijuana      3/1/2017   
3                          OH                     Franklin      6/4/2018   
4                          ON                      Toronto      5/1/2015   

  DEADLINE_DATE PROJECT_CURRENCY GOAL_IN_ORIGINAL_CURRENCY  \
0    10/10/2016              USD               100,000,000   
1     2/14/2020              USD               100,000,000   
2     3/22/2017              MXN               100,000,000   
3      7/5/2018              USD               100,000,000   
4     6/30/2015              CAD               100,000,000   

  PLEDGED_IN_ORIGINAL_CURRENCY   GOAL_IN_USD PLEDGED_IN_USD BACKERS_COUNT  \
0                            0  $100,000,000             $0             0   
1                           85  $100,000,000            $85             4   
2                           10    $5,219,374             $1             1   
3                            1  $100,000,000             $1             1   
4                            0   $80,610,122             $0             0   

      STATE  
0  canceled  
1  canceled  
2    failed  
3  canceled  
4    failed

In [160]:
#let's format and rename the columns so it doesnt look like we are yelling

ks_copy.columns = ks_copy.columns.str.capitalize()

# Some of the column names are too long
new_features = {'Caseid': 'Case_ID', 'Project_page_location_name': 'Project_Country',
                 'Project_page_location_state': 'Project_State', 'Project_page_location_county': 'City',
                 'Launched_date': 'Launched', 'Deadline_date': 'Deadline',
                 'Project_currency': 'Project_Currency', 'Goal_in_original_currency': 'Goal',
                 'Pledged_in_original_currency': 'Pledged', 'Goal_in_usd': 'Goal (USD)',
                 'Pledged_in_usd': 'Pledged (USD)', 'Backers_count': 'Backers (#)',
                 'State': 'Status'}


In [161]:
ks_copy.rename(columns=new_features, inplace=True)

In [47]:
ks_copy.sample(10)

Case_ID      Category       Subcategory Project_Country  \
479369  479370         Music             Music     Fowlerville   
290572  290573    Publishing  Children's Books          Dallas   
220965  220966           Art               Art     Summerville   
418359  418360         Music             Music   Muscle Shoals   
451771  451772  Film & Video      Film & Video       Leicester   
461392  461393           Art               Art          España   
122665  122666    Technology        Technology       Lanzarote   
400049  400050         Music             Metal   Winston-Salem   
62593    62594           Art               Art           Arabi   
433318  433319           Art    Conceptual Art         Amherst   

         Project_State                City    Launched    Deadline  \
479369              MI          Livingston    5/5/2014   5/19/2014   
290572              TX              Dallas    1/4/2020    2/3/2020   
220965              SC          Dorchester  12/16/2016   1/15/2017   
418359              AL             Colbert    6/4/2014    7/4/2014   
451771         England      Leicestershire   4/29/2014   6/28/2014   
461392          Madrid              Madrid  11/13/2019  12/13/2019   
122665  Canary Islands              Palmas    4/8/2016    5/8/2016   
400049              NC             Forsyth   1/25/2017   3/11/2017   
62593               LA  St. Bernard Parish   9/13/2016  10/15/2016   
433318              MA           Hampshire   1/14/2012   2/15/2012   

       Project_Currency    Goal Pledged Goal (USD) Pledged (USD) Backers (#)  \
479369              USD     350     464       $350          $464          10   
290572              USD   4,800   4,848     $4,800        $4,848          44   
220965              USD   7,500       0     $7,500            $0           0   
418359              USD   1,000   1,160     $1,000        $1,160          23   
451771              GBP     600     153     $1,022          $261           9   
461392              EUR     500   5,026       $559        $5,615         112   
122665              EUR  18,000     217    $20,529          $247           6   
400049              USD   1,500   1,886     $1,500        $1,887          38   
62593               USD  38,000   1,767    $38,000        $1,767          10   
433318              USD   1,000     142     $1,000          $142           5   

            Status  
479369  successful  
290572  successful  
220965      failed  
418359  successful  
451771      failed  
461392  successful  
122665      failed  
400049  successful  
62593       failed  
433318      failed

In [162]:
# format the numerical columns - Goal, Pledged, Goal (USD), Pledged (USD), Backers (#)

numeric_columns = ['Goal', 'Pledged', 'Goal (USD)', 'Pledged (USD)', 'Backers (#)']


#need to change these functions
def replace_characters(dataframe, columns, characters, replacement, regex):
    for column in columns:
        dataframe[column].replace(characters, replacement, regex=regex, inplace=True)
    return df_head(dataframe)


#change data types
def change_data_type(dataframe, columns, new_type):
   for column in columns:
       dataframe[column] = dataframe[column].astype(int)
   return df_info(dataframe)


In [163]:
replace_characters(ks_copy, numeric_columns, '\$|,|\\s', '', True)

Case_ID      Category      Subcategory Project_Country    Project_State  \
0       1  Film & Video  Science Fiction          London          England   
1       2  Film & Video          Fantasy     Los Angeles               CA   
2       3    Technology         Software          Mexico  Baja California   
3       4    Publishing       Publishing        Columbus               OH   
4       5           Art     Illustration         Toronto               ON   

             City    Launched    Deadline Project_Currency       Goal Pledged  \
0  Greater London   8/11/2016  10/10/2016              USD  100000000       0   
1     Los Angeles  12/19/2019   2/14/2020              USD  100000000      85   
2         Tijuana    3/1/2017   3/22/2017              MXN  100000000      10   
3        Franklin    6/4/2018    7/5/2018              USD  100000000       1   
4         Toronto    5/1/2015   6/30/2015              CAD  100000000       0   

  Goal (USD) Pledged (USD) Backers (#)    Status  
0  100000000             0           0  canceled  
1  100000000            85           4  canceled  
2    5219374             1           1    failed  
3  100000000             1           1  canceled  
4   80610122             0           0    failed

In [164]:
#columns where values are spaces...Checked individually prior to

# ks_copy[(ks_copy['Project_Country'] == ' ') | (ks_copy['Project_State'] == ' ') | (ks_copy['City'] == ' ')]

space_columns = ['Project_Country', 'Project_State', 'City']

replace_characters(ks_copy, space_columns, ' ', np.nan, False)

#Figure out how to replace ' ' only not words with spaces... might have to go column by column

Case_ID      Category      Subcategory Project_Country    Project_State  \
0       1  Film & Video  Science Fiction          London          England   
1       2  Film & Video          Fantasy     Los Angeles               CA   
2       3    Technology         Software          Mexico  Baja California   
3       4    Publishing       Publishing        Columbus               OH   
4       5           Art     Illustration         Toronto               ON   

             City    Launched    Deadline Project_Currency       Goal Pledged  \
0  Greater London   8/11/2016  10/10/2016              USD  100000000       0   
1     Los Angeles  12/19/2019   2/14/2020              USD  100000000      85   
2         Tijuana    3/1/2017   3/22/2017              MXN  100000000      10   
3        Franklin    6/4/2018    7/5/2018              USD  100000000       1   
4         Toronto    5/1/2015   6/30/2015              CAD  100000000       0   

  Goal (USD) Pledged (USD) Backers (#)    Status  
0  100000000             0           0  canceled  
1  100000000            85           4  canceled  
2    5219374             1           1    failed  
3  100000000             1           1  canceled  
4   80610122             0           0    failed

In [165]:
#check to see if spaces/empty still exist
ks_copy[(ks_copy['Project_Country'] == ' ') | (ks_copy['Project_State'] == ' ') | (ks_copy['City'] == ' ')]

Empty DataFrame
Columns: [Case_ID, Category, Subcategory, Project_Country, Project_State, City, Launched, Deadline, Project_Currency, Goal, Pledged, Goal (USD), Pledged (USD), Backers (#), Status]
Index: []

Empty DataFrame
Columns: [Case_ID, Category, Subcategory, Project_Country, Project_State, City, Launched, Deadline, Project_Currency, Goal, Pledged, Goal (USD), Pledged (USD), Backers (#), Status]
Index: []

In [177]:
# Backers has some empty strings - replace with 0

replace_characters(ks_copy, ['Backers (#)'], '', np.nan, False)

Case_ID      Category      Subcategory Project_Country    Project_State  \
0       1  Film & Video  Science Fiction          London          England   
1       2  Film & Video          Fantasy     Los Angeles               CA   
2       3    Technology         Software          Mexico  Baja California   
3       4    Publishing       Publishing        Columbus               OH   
4       5           Art     Illustration         Toronto               ON   

             City    Launched    Deadline Project_Currency       Goal Pledged  \
0  Greater London   8/11/2016  10/10/2016              USD  100000000       0   
1     Los Angeles  12/19/2019   2/14/2020              USD  100000000      85   
2         Tijuana    3/1/2017   3/22/2017              MXN  100000000      10   
3        Franklin    6/4/2018    7/5/2018              USD  100000000       1   
4         Toronto    5/1/2015   6/30/2015              CAD  100000000       0   

  Goal (USD) Pledged (USD) Backers (#)    Status  
0  100000000             0           0  canceled  
1  100000000            85           4  canceled  
2    5219374             1           1    failed  
3  100000000             1           1  canceled  
4   80610122             0           0    failed

In [178]:
check_nulls(ks_copy)

Here are the number of nulls per column: 
Case_ID                 0
Category                0
Subcategory             0
Project_Country      1873
Project_State        2014
City                64082
Launched                0
Deadline                0
Project_Currency        0
Goal                    0
Pledged                 0
Goal (USD)              0
Pledged (USD)           0
Backers (#)         10148
Status                  0
dtype: int64
Here are the percent of nulls per column: 
Case_ID              0.00
Category             0.00
Subcategory          0.00
Project_Country      0.37
Project_State        0.40
City                12.66
Launched             0.00
Deadline             0.00
Project_Currency     0.00
Goal                 0.00
Pledged              0.00
Goal (USD)           0.00
Pledged (USD)        0.00
Backers (#)          2.00
Status               0.00
dtype: float64


Nulls are present in Country, State, City, Backers. City has a null frequency of 12.66, but the others are 2% and lower. Will decide how to proceed later.

In [179]:
#change Status values to title

ks_copy['Status'] = ks_copy['Status'].str.title()

In [180]:
df_head(ks_copy)

Case_ID      Category      Subcategory Project_Country    Project_State  \
0       1  Film & Video  Science Fiction          London          England   
1       2  Film & Video          Fantasy     Los Angeles               CA   
2       3    Technology         Software          Mexico  Baja California   
3       4    Publishing       Publishing        Columbus               OH   
4       5           Art     Illustration         Toronto               ON   

             City    Launched    Deadline Project_Currency       Goal Pledged  \
0  Greater London   8/11/2016  10/10/2016              USD  100000000       0   
1     Los Angeles  12/19/2019   2/14/2020              USD  100000000      85   
2         Tijuana    3/1/2017   3/22/2017              MXN  100000000      10   
3        Franklin    6/4/2018    7/5/2018              USD  100000000       1   
4         Toronto    5/1/2015   6/30/2015              CAD  100000000       0   

  Goal (USD) Pledged (USD) Backers (#)    Status  
0  100000000             0           0  Canceled  
1  100000000            85           4  Canceled  
2    5219374             1           1    Failed  
3  100000000             1           1  Canceled  
4   80610122             0           0    Failed